mtp_api/apps/prison/forms.py

```diff
@@ -1,6 +1,82 @@
+import csv
+import decimal
+import io
+
 from django import forms
+from django.contrib.admin.widgets import AdminFileWidget
+from django.db import transaction
+from django.forms.models import ModelChoiceField
 from django.utils.translation import gettext_lazy as _
 
+from prison.models import Prison, PrisonerBalance, validate_prisoner_number
+
 
 class LoadOffendersForm(forms.Form):
     modified_only = forms.BooleanField(label=_('Only load updated offender locations'), required=False)
+
+
+class PrisonerBalanceUploadForm(forms.Form):
+    csv_file = forms.FileField(label=_('Prisoner balances file'), allow_empty_file=True, widget=AdminFileWidget)
+    prison = ModelChoiceField(queryset=Prison.objects.filter(use_nomis_for_balances=False))
+
+    def clean_csv_file(self):
+        csv_file = self.cleaned_data.get('csv_file')
+        if csv_file:
+            try:
+                return self.parse_balances(csv_file)
+            except Exception as e:
+                self.add_error('csv_file', f'Could not parse file: {e}')
+        return None
+
+    @classmethod
+    def parse_balances(cls, csv_file):
+        csv_contents = csv_file.read().decode()
```

Comment:
Are we using a custom `django.core.files.storage.Storage` subclass, or is `csv_file` just be a standard Python file object right? If it's the latter can't you just do `csv_reader = csv.DictReader(csv_file)`?

Reply:
we tried that but the csv reader wants text and we couldn't get it to re-open the file in text mode (the special `file`-like object opens in binary mode)